src/pyduke_energy/realtime.py

```
@@ -0,0 +1,244 @@
+            The paho-mqtt client
+        userdata
+            user data passed by the client
+        mid : int
+            message id#
+
+        This will call the client.disconnect() method
+        """
+        _LOGGER.debug("MQTT unsubscribed msg_id: %s", str(mid))
+        client.disconnect()
+
+    def on_discon(self, _client: mqtt.Client, _userdata, conn_res):
+        """On Disconnect callback.
+
+        Parameters
+        ----------
+        client : mqtt.Client
+            The paho-mqtt client
+        userdata
+            user data passed by the client
+        conn_res
+            Disconnect error code
+        """
+        if conn_res:
+            _LOGGER.error(
+                "MQTT disconnect error, result code: %s", mqtt.error_string(conn_res)
+            )
+        else:
+            _LOGGER.debug(
+                "MQTT disconnected with result code: %s", mqtt.error_string(conn_res)
+            )
+        self.disconnected.set_result(conn_res)
+
+    def on_msg(self, msg):
+        """On Message Callback
+
+        Parameters
+        ----------
+        msg : MQTTMessage
+            This is a class with members topic, payload, qos, retain
+        """
+        _LOGGER.debug("rx msg on %s\n%s", msg.topic, msg.payload.decode("utf8"))
+
+    def _on_msg(self, _client: mqtt.Client, _userdata, msg):
+        """Private On Message callback.
+
+        Parameters
+        ----------
+        client : mqtt.Client
+            The paho-mqtt client
+        userdata
+            user data passed by the client
+        msg : MQTTMessage
+            This is a class with members topic, payload, qos, retain
+        """
+        # _LOGGER.debug("rx msg on %s\n%s", msg.topic, msg.payload.decode('utf8'))
+        # _LOGGER.debug("_on_msg")
+        if not self.rx_msg:
+            _LOGGER.warning("Unexpected message: %s", msg)
+        else:
+            self.rx_msg.set_result((msg.payload.decode("utf8")))
+            self.on_msg(msg)
```

maybe check on_msg is not None here in case no callback was implemented. since that seems like a weird case, you could raise an Exception

---

My intention was for there to be a "default" on_msg callback in the class, in this case it just prints to the logger. But then if you want to subclass the DukeEnergyRealtime ( I think thats the right term) and define a more specialized version of on_msg it would allow you to do so. Does that make sense? is that a good way to accomplish this? It still doesn't feel ideal to me since the on_msg callback is blocking. Again, I'm very new to python, so i might be missing something obvious.

---

I think the approach is good. So what I'm pointing out is that in theory, someone could just not implement `on_msg` or even use the `DukeEnergyRealtime` class without subclassing it. If you want this to be allowed (it sounds like you do), then you will have to consider the case that self.on_msg doesn't exist, which means the line here will throw an error.

Something like this will handle it

```
if self.on_msg:
    self.on_msg(msg)
```

If you don't want this to be a possibility, then you'd be best served marking the `DukeEnergyRealtime` and the method `on_msg` as abstract (https://stackoverflow.com/a/13646263).